File: app/src/main/java/com/dharmendra/intermediatekotlin/todo/TodoView.java
package com.dharmendra.intermediatekotlin.todo;

import android.content.Context;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.dharmendra.intermediatekotlin.R;
import com.dharmendra.intermediatekotlin.Todo;

public class TodoView extends ConstraintLayout {
    private CheckBox completeCheckBox;
    private TextView descriptionView;

    public TodoView(@NonNull Context context) {
        super(context);
    }

    public TodoView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TodoView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void initView(Todo todo){
        completeCheckBox = findViewById(R.id.completeCheckBox);
        descriptionView  = findViewById(R.id.descriptionView);

        descriptionView.setText(todo.getDescription());
        completeCheckBox.setChecked(todo.isComplete());

        if (todo.isComplete()){
            createStrikeThrough();
        }
        setUpCheckStateListener();
    }

    private void setUpCheckStateListener(){
        completeCheckBox.setOnCheckedChangeListener((button,isChecked)->{
            if (isChecked)
                createStrikeThrough();
            else
                removeStrikeThrough();
        });
    }

    private void createStrikeThrough(){
        descriptionView.setPaintFlags(descriptionView.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
    }

    private void removeStrikeThrough(){
        descriptionView.setPaintFlags(descriptionView.getPaintFlags() & ~(Paint.STRIKE_THRU_TEXT_FLAG));
    }




}

File: app/src/main/java/com/dharmendra/intermediatekotlin/todo/TodoView.kt
package com.dharmendra.intermediatekotlin.todo

import android.content.ComponentCallbacks
import android.content.Context
import android.graphics.Paint
import android.util.AttributeSet

import androidx.constraintlayout.widget.ConstraintLayout

import android.widget.CompoundButton
import com.dharmendra.intermediatekotlin.Todo
import com.dharmendra.intermediatekotlin.extfunction.removeStrikeThrough
import com.dharmendra.intermediatekotlin.extfunction.setStrikeThrough
import kotlinx.android.synthetic.main.view_todo.view.*

class TodoView @JvmOverloads constructor(context: Context,
attrs: AttributeSet?=null,
defStyleAttr: Int=1)  : ConstraintLayout(context,attrs,defStyleAttr) {

    fun initView(todo: Todo,callbacks: ((Boolean) -> Unit)?=null) {
        descriptionView.setText(todo.description)
        completeCheckBox.setChecked(todo.isComplete)
        if (todo.isComplete) {
            descriptionView.setStrikeThrough()
        }
        setUpCheckStateListener(todo,callbacks )
    }

    private fun setUpCheckStateListener(todo:Todo,callbacks: ((Boolean)->Unit)?=null) {

        completeCheckBox.setOnCheckedChangeListener {
               _, isChecked->
            todo.isComplete = isChecked
            callbacks?.invoke(isChecked)
            if (isChecked)
                descriptionView.setStrikeThrough()
            else
                descriptionView.removeStrikeThrough()
        }
    }

//    private fun createStrikeThrough() {
//        descriptionView.apply {
//            paintFlags = paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
//        }
//    }

//    private fun removeStrikeThrough() {
//        descriptionView.apply {
//            paintFlags = paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
//
//        }
//    }
}